src/View/Helper/FormHelper.php

```diff
@@ -2488,28 +2494,41 @@ public function getValueSources(): array
     }
 
     /**
-     * Extracts valid value sources.
+     * Validate value sources.
      *
      * @param string[] $sources A list of strings identifying a source.
-     * @return string[]
+     * @return void
+     * @throws \InvalidArgumentException If sources list contains invalid value.
      */
-    protected function extractValidValueSources(array $sources): array
+    protected function validateValueSources(array $sources): void
     {
-        return array_values(array_intersect($sources, $this->supportedValueSources));
+        $diff = array_diff($sources, $this->supportedValueSources);
+
+        if ($diff) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid value source(s): %s. Valid values are: %s',
+                implode(', ', $diff),
+                implode(', ', $this->supportedValueSources)
+            ));
+        }
     }
 
     /**
      * Sets the value sources.
      *
-     * Valid values are set in `$supportedValueSources`.
-     * You need to supply one valid context or multiple, as a list of strings. Order sets priority.
+     * You need to supply one or more valid source, as a list of strings.
+     * Order sets priority.
      *
+     * @see FormHelper::$supportedValueSources for valid values.
      * @param string|string[] $sources A string or a list of strings identifying a source.
      * @return $this
      */
```

Comment:
```suggestion
     * @throws \InvalidArgumentException If sources list contains invalid value.
     */
```

Isn't this the public api for users?